### PR TITLE
Allow automatic fetching of secrets when provisioning

### LINKED ISSUE #519: Allow provisiong more than a server at the same time
## What? Why?

This is a known old thing but it came up again in https://github.com/openfoodfoundation/ofn-install/pull/518.

While we are able to deploy a bunch of servers together, we can't do the same when provisioning resulting in a waste of time in manual work, that can always lead to errors.

This is because currently, due the structure of the `ofn-secrets` repo the secrets file needs to be provided as follows for each individual server.

```
ansible-playbook playbooks/provision.yml --limit=es-staging -e "@../ofn-secrets/es-staging/secrets.yml"
```

Otherwise, there is no way to tell Ansible which one of the matching variables in the various `secrets.yml` needs to end up in each server. Aka. `fr-prod` might end up with the UK's Stripe API key.

## Potential solution

@danypr92 suggests we follow Coopdevs' convention of having the secrets file next to the config.yml of each host_var. See https://gitlab.com/coopdevs/odoo-test-provisioning/tree/master/inventory/host_vars/odoo-test.coopdevs.org

This is almost the same we already have but simpler. Given the folder structure maps one to one, instead of a separate repo we merge them into a single one, which is always simpler to manage. for instance: do you always pull changes in `ofn-secrets` when running the provisioning against a server?

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
Closes #519

We've had to provision all servers a couple of times recently, and we need to do it again pretty soon.

This PR enables automatic fetching and updating of local secrets files. The encrypted secrets files are stored locally under `ofn-install/inventory/host_vars/<host>/secrets.yml` alongside the `config.yml` files. These are then automatically applied to each host when running a playbook, and are applied correctly when using groups such as `--limit europe`. This can be done manually, or can be set to run every time the `provision.yml` playbook is run.

These changes are entirely optional, and will require use of an SSH key for your github account when fetching secrets from the private repo. We can now use: `ansible-playbook playbooks/provision.yml --limit all-prod` :tada:

## Usage

**Update local secrets:** `ansible-playbook playbooks/fetch_secrets.yml` (no --limit)

**Update local secrets and provision:** `ansible-playbook playbooks/provision.yml --limit uk-staging -e "fetch_ofn_secrets=TRUE"`

**Automatically fetch the latest secrets every time before provisioning:**
- Set this line in your `~/.bashrc` file: `export FETCH_OFN_SECRETS="TRUE"`
- Pick up the bashrc changes with `source ~/.bashrc` (first time only)
- Run the provision playbook as normal, eg `ansible-playbook playbooks/provision.yml --limit uk-staging`

### CODE REVIEW

**Comment by Matt-Yorkley** [comment]
> this deserves some update on some wiki page for users, right?

Yes! It's only usable by users that have access to the secrets repo though.

**Comment by sauloperez** [comment]
Cool! This will make us much more efficient! would you mind adding that role for deployment as well? This way we can close https://github.com/openfoodfoundation/ofn-install/issues/492 too

**Comment by Matt-Yorkley** [comment]
> would you mind adding that role for deployment as well?

We don't need it for deployment, right?

**Comment by Matt-Yorkley** [comment]
Ah, I've just read #492 ...

I'm pretty reluctant to add secrets management back into deployments to be honest...

**Comment by sauloperez** [comment]
but then how do we notify DD or any other service?

**Comment by Matt-Yorkley** [comment]
I guess we could add it for deploying manually to production, but it'll be more complicated if we want it for Semaphore deployments to staging servers...

**Comment by sauloperez** [comment]
yes, that's the only use case we actually care about

**Comment by Matt-Yorkley** [comment]
Hmmm... the datadog API keys don't really get updated very often right? And that's the only real secret that will be used in deployment? I think once we  start using this PR, there will be copies of the secrets files already in place locally for any core dev/sysadmin that uses ofn-install, so the secrets will be there anyway when deploying.

**Comment by sauloperez** [comment]
ok, I buy it...

**Comment by sauloperez** [comment]
Unfortunately, it failed on `Install certbot`

```
TASK [coopdevs.certbot_nginx : Install certbot] **************************************************************************************************************
Wednesday 18 September 2019  12:36:01 +0200 (0:00:05.946)       0:13:24.826 ***
fatal: [staging.openfoodnetwork.org.uk]: FAILED! => {"cache_update_time": 1568802346, "cache_updated": false, "changed": false, "msg": "'/usr/bin/apt-get -y -o \"Dpkg::Options::=--force-confdef\" -o \"Dpkg::Options::=--force-confold\"     install 'certbot=0.28.0-1+ubuntu16.04.1+certbot+4'' failed: E: Version '0.28.0-1+ubuntu16.04.1+certbot+4' for 'certbot' was not found\n", "rc": 100, "stderr": "E: Version '0.28.0-1+ubuntu16.04.1+certbot+4' for 'certbot' was not found\n", "stderr_lines": ["E: Version '0.28.0-1+ubuntu16.04.1+certbot+4' for 'certbot' was not found"], "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state information...\n", "stdout_lines": ["Reading package lists...", "Building dependency tree...", "Reading state information..."]}
fatal: [staging.katuma.org]: FAILED! => {"cache_update_time": 1568802348, "cache_updated": false, "changed": false, "msg": "'/usr/bin/apt-get -y -o \"Dpkg::Options::=--force-confdef\" -o \"Dpkg::Options::=--force-confold\"     install 'certbot=0.28.0-1+ubuntu16.04.1+certbot+4'' failed: E: Version '0.28.0-1+ubuntu16.04.1+certbot+4' for 'certbot' was not found\n", "rc": 100, "stderr": "E: Version '0.28.0-1+ubuntu16.04.1+certbot+4' for 'certbot' was not found\n", "stderr_lines": ["E: Version '0.28.0-1+ubuntu16.04.1+certbot+4' for 'certbot' was not found"], "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state information...\n", "stdout_lines": ["Reading package lists...", "Building dependency tree...", "Reading state information..."]}
fatal: [staging.openfoodnetwork.org.au]: FAILED! => {"cache_update_time": 1568802352, "cache_updated": false, "changed": false, "msg": "'/usr/bin/apt-get -y -o \"Dpkg::Options::=--force-confdef\" -o \"Dpkg::Options::=--force-confold\"     install 'certbot=0.28.0-1+ubuntu16.04.1+certbot+4'' failed: E: Version '0.28.0-1+ubuntu16.04.1+certbot+4' for 'certbot' was not found\n", "rc": 100, "stderr": "E: Version '0.28.0-1+ubuntu16.04.1+certbot+4' for 'certbot' was not found\n", "stderr_lines": ["E: Version '0.28.0-1+ubuntu16.04.1+certbot+4' for 'certbot' was not found"], "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state information...\n", "stdout_lines": ["Reading package lists...", "Building dependency tree...", "Reading state information..."]}

RUNNING HANDLER [restart postgres] ***************************************************************************************************************************
Wednesday 18 September 2019  12:36:08 +0200 (0:00:06.974)       0:13:31.801 ***
	to retry, use: --limit @/Users/pau/dev/ofn-install/playbooks/provision.retry

PLAY RECAP ***************************************************************************************************************************************************
staging.katuma.org         : ok=117  changed=4    unreachable=0    failed=1
staging.openfoodfrance.org : ok=5    changed=0    unreachable=0    failed=1
staging.openfoodnetwork.org.au : ok=81   changed=2    unreachable=0    failed=1
staging.openfoodnetwork.org.uk : ok=79   changed=3    unreachable=0    failed=1
```

File: roles/fetch_secrets/tasks/update_secrets.yml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+---
+
+- name: extract host_id from filepath
+  set_fact:
+    secrets_host_id: "{{ secrets_host_path.path.split('/tmp/ofn-secrets/')[1] }}"
+
+- name: copy {{ secrets_host_id }} secrets file to relevent host_vars directory
```
Comment (enricostano) [inline roles/fetch_secrets/tasks/update_secrets.yml]: `s/relevent/relevant/`

**Comment by sauloperez** [comment]
That error was due to my setup. I didn't reinstall dependencies.

Tested and working

```
PLAY RECAP ****************************************************************************************************************************************************************************
staging.katuma.org         : ok=146  changed=8    unreachable=0    failed=0   
staging.openfoodfrance.org : ok=124  changed=19   unreachable=0    failed=0   
staging.openfoodnetwork.org.au : ok=162  changed=5    unreachable=0    failed=0   
staging.openfoodnetwork.org.uk : ok=135  changed=1    unreachable=0    failed=0   

Wednesday 18 September 2019  14:03:23 +0200 (0:00:03.530)       0:20:17.250 *** 
=============================================================================== 
```

**Comment by sauloperez** [comment]
CI is failing due to the ruby version which in the app is now pointing to 2.1.9 but this PR branch doesn't support it. I know it's bad practice but I'll merge and we'll fix it today or tomorrow.

**Comment by Matt-Yorkley** [comment]
2.1.9 issues in the ofn-install builds are now gone after the Ruby bump has been merged. :heavy_check_mark:

**Pull request merged**